src/Common.mo

```diff
@@ -80,35 +83,74 @@ module {
       Nat32.fromIntWrap(Nat8.toNat(bytes[offset + 0]));
   };
 
+  // Write given value as 16-bit little endian into array starting at offset.
+  public func writeLE16(bytes : [var Nat8], offset : Nat, value : Nat16) {
+    let first : Nat8 = Nat8.fromIntWrap(Nat16.toNat(value));
+    let second : Nat8 = Nat8.fromIntWrap(Nat16.toNat(value >> 8));
+
+    bytes[offset] := first;
+    bytes[offset + 1] := second;
+  };
+
   // Write given value as 32-bit little endian into array starting at offset.
   public func writeLE32(bytes : [var Nat8], offset : Nat, value : Nat32) {
-    bytes[offset + 3] := Nat8.fromNat(Nat32.toNat((value & 0xFF000000) >> 24));
-    bytes[offset + 2] := Nat8.fromNat(Nat32.toNat((value & 0xFF0000) >> 16));
-    bytes[offset + 1] := Nat8.fromNat(Nat32.toNat((value & 0xFF00) >> 8));
-    bytes[offset] := Nat8.fromNat(Nat32.toNat((value & 0xFF)));
+    let first : Nat16 = Nat16.fromIntWrap(Nat32.toNat(value));
+    let second : Nat16 = Nat16.fromIntWrap(Nat32.toNat(value >> 16));
+
+    writeLE16(bytes, offset, first);
+    writeLE16(bytes, offset + 2, second);
   };
 
   // Write given value as 64-bit little endian into array starting at offset.
   public func writeLE64(bytes : [var Nat8], offset : Nat, value : Nat64) {
-    bytes[offset + 7] :=
-      Nat8.fromNat(Nat64.toNat((value & 0xFF00000000000000) >> 56));
-    bytes[offset + 6] :=
-      Nat8.fromNat(Nat64.toNat((value & 0xFF000000000000) >> 48));
-    bytes[offset + 5] :=
-      Nat8.fromNat(Nat64.toNat((value & 0xFF0000000000) >> 40));
-    bytes[offset + 4] :=
-      Nat8.fromNat(Nat64.toNat((value & 0xFF00000000) >> 32));
-    bytes[offset + 3] := Nat8.fromNat(Nat64.toNat((value & 0xFF000000) >> 24));
-    bytes[offset + 2] := Nat8.fromNat(Nat64.toNat((value & 0xFF0000) >> 16));
-    bytes[offset + 1] := Nat8.fromNat(Nat64.toNat((value & 0xFF00) >> 8));
-    bytes[offset] := Nat8.fromNat(Nat64.toNat((value & 0xFF)));
+    let first : Nat32 = Nat32.fromIntWrap(Nat64.toNat(value));
+    let second : Nat32 = Nat32.fromIntWrap(Nat64.toNat(value >> 32));
+
+    writeLE32(bytes, offset, first);
+    writeLE32(bytes, offset + 4, second);
   };
 
   // Copy data from src into dest from/at the given offsets.
   public func copy(dest: [var Nat8], destOffset : Nat, src: [Nat8],
     srcOffset : Nat, count : Nat) {
       for (i in Iter.range(0, count - 1)) {
         dest[destOffset + i] := src[srcOffset + i];
+    };
+  };
+
+  // Parses given text as unsigned integer, returns null if it contains
+  // non-number characters.
+  public func textToNat(input : Text) : ?Nat {
+    var result : Nat = 0;
+    for (asciiVal in Text.encodeUtf8(input).vals()) {
+      if (asciiVal < 0x30 or asciiVal > 0x39) {
+        return null;
       };
+
+      result *= 10;
+      result += Nat8.toNat((asciiVal - 48));
     };
+
+    return ?result;
+  };
+
+  public func encodeVarint(value : Nat) : [Nat8] {
+    assert(value < 0x10000000000000000);
+
+    return if (value < 0xfd) {
```

Comment:
Could you please also add some in-line documentation? For example, why `value < 0xfd`?